### PR TITLE
Feature to let host prefetch the snapshot and then use that to load from in odsp driver

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Feature to let host prefetch the snapshot and then use that to load from in odsp driver. Host can initiate a prefetch call for the snapshot and then use the cache to pass to the driver. Driver will try to extract the result from the snapshot prefetch cache and use it for the load.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/index.ts [hunk 1]
```diff
@@ -50,3 +50,5 @@ export {
 } from "./odspFluidFileLink";
 
 export { parseCompactSnapshotResponse } from "./compactSnapshotParser";
+
+export { SnapshotPrefetchResultCache } from "./odspCache";
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/index.ts:54]: why do we need to export it from the driver package?

File: packages/drivers/odsp-driver/src/odspCache.ts [hunk 2]
```diff
@@ -105,6 +107,8 @@ export interface INonPersistentCache {
 	 * Cache of resolved/resolving file URLs
 	 */
 	readonly fileUrlCache: PromiseCache<string, IOdspResolvedUrl>;
+
+	readonly snapshotPrefetchResultCache?: SnapshotPrefetchResultCache;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspCache.ts:111]: maybe add comment on the semantics and how it's used?

File: packages/drivers/odsp-driver/src/odspCache.ts [hunk 3]
```diff
@@ -118,10 +122,29 @@ export interface IOdspCache extends INonPersistentCache {
 }
 
 export class NonPersistentCache implements INonPersistentCache {
+	constructor(readonly snapshotPrefetchResultCache?: SnapshotPrefetchResultCache) {
+		if (this.snapshotPrefetchResultCache === undefined) {
+			this.snapshotPrefetchResultCache = new PromiseCache<
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspCache.ts:127]: Curious, can you do here new SnapshotPrefetchResultCache() ?

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts [hunk 4]
```diff
@@ -206,15 +211,17 @@ export class OdspDocumentServiceFactoryCore implements IDocumentServiceFactory {
 	 * @param getWebsocketToken - function that can provide a token for accessing the web socket. This is also
 	 * to as the "Push" token in SPO. If undefined then websocket token is expected to be returned with joinSession
 	 * response payload.
-	 * @param storageFetchWrapper - if not provided FetchWrapper will be used
-	 * @param deltasFetchWrapper - if not provided FetchWrapper will be used
 	 * @param persistedCache - PersistedCache provided by host for use in this session.
+	 * @param hostPolicy - Policy for storage provided by host.
+	 * @param snapshotPrefetchResultCache - Non Persistant cache which could contain snapshot promise if it has been
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:216]: I'm not sure about this part. I think you are trying to solve more generic problem of - external code controlling lifetime of this cache, and allowing share it across multiple instances of driver factories. I'm not sure consumers do need the latter, and I do not think they would be happy about the former.
I wonder if
1. prefetchLatestSnapshot() should take an instance of the driver factory and leverage that to get access to non-persistent cache
2. consumers need to use same driver factory instance when doing prefetch() calls and loading.

It's hard to say which direction is right, but I think we need to present these two options (your current approach and what I've described) to Loop and figure out which one is easier to work with.

Can you please follow up here?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 5]
```diff
@@ -235,21 +239,56 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
...
+						| undefined;
+
+					let method: string | undefined;
+					const snapshotCacheKey = createCacheSnapshotKey(this.odspResolvedUrl);
+					let prefetchedSnapshotP:
+						| Promise<ISnapshotContentsWithEpoch | undefined>
+						| undefined;
+					// Only use the prefetched snapshot from non persistent cache in the load flow.
+					if (this.firstVersionCall) {
+						prefetchedSnapshotP = this.cache.snapshotPrefetchResultCache
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:254]: why does it have to be under IF(this.firstVersionCall) branch?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 6]
```diff
@@ -289,47 +330,81 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
...
-							retrievedSnapshot = promiseRaceWinner.value;
-							method = promiseRaceWinner.index === 0 ? "cache" : "network";
-
-							if (retrievedSnapshot === undefined) {
-								// if network failed -> wait for cache ( then return network failure)
-								// If cache returned empty or failed -> wait for network (success of failure)
-								if (promiseRaceWinner.index === 1) {
-									retrievedSnapshot = await cachedSnapshotP;
-									method = "cache";
+							if (prefetchedSnapshotP) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:337]: I got lost here on why we need to much change :) It feels like we should simply treat prefetchedSnapshotP and networkSnapshotP the same way. I.e. I do not think there is that much value to record differences between "network" and "prefetched" - after all the difference is only that network call started a bit sooner. If you really want to capture that, I'd rather record (in non-persistent cache) when it started, and record delta as a property on ObtainSnapshot event.
The rest of the logic seems like should not change at all.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 7]
```diff
@@ -111,7 +126,15 @@ export async function prefetchLatestSnapshot(
...
+			)
+				.then((value) => {
+					assert(!!snapshotEpoch, "prefetched snapshot should have a valid epoch");
+					snapshotContentsWithEpochP.resolve({ ...value, fluidEpoch: snapshotEpoch });
+				})
+				.catch((err) => {
+					snapshotContentsWithEpochP.reject(err);
+					throw err;
+				});
 			assert(cacheP !== undefined, 0x1e7 /* "caching was not performed!" */);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:138]: I believe you need to remove cache line from cache after request is done.
If you do not do so, then all future loads will keep using that cache line.
Instead, we should rely on persistent cache (after load finished), and overall load flow of continuously (on each load) refreshing permanent cache.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:138]: @vladsud What do you mean by removing cache line?

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:138]: You mean once the response is stored in permanent cache, then we should remove it from the non persistent cache?

File: packages/drivers/odsp-driver/src/odspCache.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspCache.ts:127]: Since now, I am sharing the driver factory as you suggested and cosmin also recommended, we don't need it. So removing.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspCache.ts:127]: But you can do it what you were asking.

File: packages/drivers/odsp-driver/src/index.ts (hunk 1 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/index.ts:54]: Removed.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts (hunk 7 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:138]: Removed the promise from the non-persistent cache, once it is stored in the persistent cache.

File: packages/drivers/odsp-driver-definitions/src/factory.ts [hunk 8]
```diff
@@ -141,4 +141,10 @@ export interface HostStoragePolicy {
 	 * where 'createLinkType' was requested.)
 	 */
 	enableSingleRequestForShareLinkWithCreate?: boolean;
+
+	/**
+	 * True if host does not want the storage service to use the prefetch cache to get the snapshot. Undefined will be treated
+	 * as false. This is if the host wants to do some A/B testing.
+	 */
+	avoidPrefetchSnapshotCache?: boolean;
```
Comment (vladsud) [inline packages/drivers/odsp-driver-definitions/src/factory.ts:149]: Is this useful setting to expose? I.e. what would be a scenario?
I feel like you are adding it to be able to disable new code in case of an issue. And if so, is it better to use feature gate (kill-bit switch) instead? It's more directly telling the intention.

File: packages/drivers/odsp-driver/package.json [hunk 9]
```diff
@@ -35,7 +35,7 @@
 		"prettier:fix": "prettier --write . --ignore-path ../../../.prettierignore",
 		"test": "npm run test:mocha",
 		"test:coverage": "nyc npm test -- --reporter xunit --reporter-option output=nyc/junit-report.xml",
-		"test:mocha": "mocha --ignore 'dist/test/types/*' --recursive dist/test -r node_modules/@fluidframework/mocha-test-setup --unhandled-rejections=strict",
+		"test:mocha": "mocha --ignore 'dist/test/types/*' --recursive dist/test -r node_modules/@fluidframework/mocha-test-setup --unhandled-rejections=strict -g xxxTests",
```
Comment (vladsud) [inline packages/drivers/odsp-driver/package.json:38]: Not sure what is that? But the name is not great :)

Comment (jatgarg) [reply packages/drivers/odsp-driver/package.json:38]: Not ready for review yet :) Want to clean some things.

Comment (jatgarg) [reply packages/drivers/odsp-driver/package.json:38]: It was just to run my specific tests for testing. The tests are not fixed yet, so it was still there.

File: packages/drivers/odsp-driver/src/odspCache.ts [hunk 10]
```diff
@@ -124,4 +132,17 @@ export class NonPersistentCache implements INonPersistentCache {
...
+		ISnapshotContentsWithEpoch
+	>();
+}
+
+export interface ISnapshotContentsWithEpoch extends ISnapshotContents {
+	fluidEpoch: string;
+}
+
+export function snapshotPrefetchCacheKeyFromEntry(entry: ICacheEntry): string {
+	return `${entry.file.docId}_${entry.type}_${entry.key}`;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspCache.ts:147]: This API is a bit weird, because it assumes that it will be called for entry.type ==- "snapshot", but it does not assert that.
And it's not very clear from API alone on what it does - one needs to look at usage to realize that it provides a mapping between persistent cache (snapshot) key to a key in prefetchCache.

I think we can be served just fine by more generic function (name) doing same thing, like existing getKeyForCacheEntry(), It would be also great to make LocalPersistentCache.keyFromEntry(), OdspPersistentCache.keyFromEntry(), OdspSampleCache.keyFromEntry() to not reimplement that same logic but rather call directly into getKeyForCacheEntry()

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts [hunk 11]
```diff
@@ -51,7 +51,7 @@ import {
 export class OdspDocumentServiceFactoryCore implements IDocumentServiceFactory {
 	public readonly protocolName = "fluid-odsp:";
 
-	private readonly nonPersistentCache = new NonPersistentCache();
+	public readonly nonPersistentCache: INonPersistentCache = new NonPersistentCache();
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:54]: public is not good. Maybe you need protected?

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 12]
```diff
@@ -111,9 +126,19 @@ export async function prefetchLatestSnapshot(
...
-			await cacheP;
+			// Remove it from the non persistent cache once it is cached in the persistent cache as then it will
+			// be fetched from persistent cache.
+			await cacheP.then(() => nonPersistentCache?.remove(nonPersistentCacheKey));
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:141]: I think this needs to be on .finally() path, not on .then() path.
I.e. we do not want to get stuck completely on first error forever. If it failes, it fails all in flight loads, but anybody can retry.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 13]
```diff
@@ -235,21 +239,61 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
...
+								this.logger.sendTelemetryEvent(
+									{
+										eventName: "PrefetchSnapshotError",
+										concurrentSnapshotFetch:
+											this.hostPolicy.concurrentSnapshotFetch,
+									},
+									err,
+								);
+								throw err;
+							}) ?? this.fetchSnapshot(
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:278]: I believe this flow breaks this.hostPolicy.concurrentSnapshotFetch === false promise.
The promise is - we will consult cache before making network call.
It feels like changes to logic should be not in this fuction, but rather be part of OdspDocumentStorageService.fetchSnapshot() flow.

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts (hunk 4 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts]: Cosmin preferred to take the approach you suggested, so made changes accordingly.

File: packages/drivers/odsp-driver/src/odspCache.ts (hunk 10 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspCache.ts:147]: done

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 6 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Changed the logic.

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts (hunk 11 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts]: I am now exposing just the snapshot cache through the getter. We won't be able to access in prefetch if we make it protected.

File: packages/drivers/odsp-driver-definitions/src/factory.ts (hunk 8 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver-definitions/src/factory.ts:149]: No, it is not for that. Cosmin mentioned that he wants to have some hostpolicy in case they want to do a good A/B testing for performance like when they call prefetch api and then load the container, then how fast is to load if we use the prefetch promise from cache and when we are not. So, they can write their logic and use this policy to achieve that. 
We don't need a feature gate because if some error occurred while prefetching then, according to current logic I am simply just proceeding with the network call according to old flow. 
Once we have more confidence then I can just throw if the prefetch fails and not make a network call.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 14]
```diff
@@ -417,6 +422,34 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
 		hostSnapshotOptions: ISnapshotOptions | undefined,
 		scenarioName?: string,
 	) {
+		// Don't look into cache, if the host specifically tells us so.
+		if (!this.hostPolicy.avoidPrefetchSnapshotCache) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: Is avoidPrefetchSnapshotCache mostly a safety net? I.e. if we have a bug, app can disable this code path?
Or do you think there is legit workflow where host wants to disable this feature?
It feels like it's the former, and if so, it feels like risk is low - they can always work-around an issue by skipping / awaiting prefetch call before launching load (slower, but avoids new code path).

I'm simply looking for simplicity - it's very hard for our customers to understand full matrix of options...

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: No, it is not for that. Cosmin mentioned that he wants to have some hostpolicy in case they want to do a good A/B testing for performance like when they call prefetch api and then load the container, then how fast is to load if we use the prefetch promise from cache and when we are not. So, they can write their logic and use this policy to achieve that.
We don't need a feature gate because if some error occurred while prefetching then, according to current logic I am simply just proceeding with the network call according to old flow.
Once we have more confidence then I can just throw if the prefetch fails and not make a network call.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: They can just avoid this option and not pass anything if they don't want to use this.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 15]
```diff
@@ -86,13 +91,15 @@ export async function prefetchLatestSnapshot(
 			storageToken,
 			odspLogger,
 			snapshotOptions,
-			snapshotFormatFetchType,
+			undefined,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:94]: Should we remove this argument from the list of arguments to prefetchLatestSnapshot() ?
That's breaking change, so maybe that should be done in next.
In general, should we remove from code ability to dictate what format to fetch? It feels like we have been in prod enough to say that we will always ask for binary, with fallback to json, and thus any code that provides the opposite should be cleaned.
I'd suggest doing it as separate PR, and across all code paths.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:94]: Yes, I was thinking of doing it in the next branch.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 16]
```diff
@@ -111,10 +127,25 @@ export async function prefetchLatestSnapshot(
...
+					snapshotContentsWithEpochP.reject(err);
+					// Remove from cache if prefetch was not successful.
+					snapshotNonPersistentCache?.remove(nonPersistentCacheKey);
+					throw err;
+				});
 			assert(cacheP !== undefined, 0x1e7 /* "caching was not performed!" */);
-			await cacheP;
+			// Remove it from the non persistent cache once it is cached in the persistent cache or an error
+			// occured while caching.
+			await cacheP.finally(() => snapshotNonPersistentCache?.remove(nonPersistentCacheKey));
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:144]: It's a bit hard to follow code.
I wonder if you could move "await cacheP" after line 133 (to successful path of prefetching), and thus do snapshotNonPersistentCache?.remove() in .finally block to fetchSnapshotWithRedeem().

I.e.

```ts
fetchSnapshotWithRedeem()
    .then(() => {
        ...
        await cacheP;
    }.catch(() => {
        ...
    }.finally (() => {
        snapshotNonPersistentCache?.remove
    }
}

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 14 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: I'd actually suggest not doing A/B testing like that. In a sense that we know that leveraging in-flight prefetch would be always faster than doing another parallel call.
So really, we are after only measuring by how much it would be faster. And we can ensure we have enough telemetry in place to measure that, without requiring A/B testing.

The only place where A/B testing would be useful if we think that (for some reason) leveraging this new code would make things worse. But I'd think that's more about reliability than perf, as I have hard time seeing where it would be slower that way. Plus, we have natural A/B testing here in the form of old vs. new driver - we will have at least 1-2 weeks where we have both in usage and can easily do this kind of comparison.

At the end of the day it's up to Loop, but I'd make sure that
- we remove this option in next - cleaner / simpler code
- we do have enough telemetry in place to answer all these questions directly.

Comment (cosmincatrinescu) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: We've always struggled with assessing the actual gains of perf improvements by looking exclusively at old vs new code sessions (code changes in these layer may be landing at different times in hosts, prefetch may have been called or not on some scenarios, etc). The reliable way to measure the impact of this change itself would be to focus on scenarios telemetry that (1) do prefetch so those sessions are legit for experimentation and (2) split those into A/B. Hence I suggested the extra flags. Seeing the gains accurately would help.

If you think old/new telemetry without the A/B flag is actually allowing for as accurate assessment, due to large volume of sessions maybe and focusing on one single scenario etc, then I'm ok with that. It's just less headache in data digging when you can A/B. But agreed, the code can look gnarly when forcing A/B flags.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: I am going to leave as is for now. It is just an if statement. We can always remove later on.

Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:426]: I agree that overall doing it via version differences is hard. But in this case, I'd argue we should be able to tell precisely (per session) what the impact of these changes is. Maybe other way to say it - we should use all the tools we have here, but I expect A/B testing data will be noisier than direct telemetry (using only new bits) on how much time was saved in a single TreesLatest call by observing the diff between when prefetch started and when Container Load path started to wait for that in-flight call. I want to make sure that we have telemetry to collect those direct measurements.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts (hunk 16 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:144]: Done

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 14 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:438]: Recording this time now too.
